

## 의사결정나무 (회귀)     
     

> sklearn.tree.DecisionTreeRegressor

- 연속형 목표변수   
    
|기준값 |분리기준|  |
|------|---|---|
|분산분석에서 F통계량|p값이 가장 작은 예측변수와 그 때의 최적분리에 의해서 자식마디를 형성|  |
|분산의 감소량|분산의 감소량을 최대화하는 기준의 최적분리에 의해서 자식마디를 형성|  |
     

* <b> criterion{“squared_error”, “friedman_mse”, “absolute_error”, “poisson”}, default=”squared_error”

    - 분할할 때 사용할 함수
* splitter:{“best”, “random”}, default=”best”
    - 분할방법
        - best: 분할 시가장 좋은 노드 선택
        - random: 임의적으로 노드 선택
* max_depth:int, default=None
    - 트리의 최대 깊이
* min_samples_split:int or float, default=2
    - 내부 노드를 분할하는 데 필요한 최소 샘플 수
* min_samples_leaf:int or float, default=1
    - 리프노드에 있어야 하는 최소 샘플 수
* max_features:int, float or {“auto”, “sqrt”, “log2”}, default=None
    - 최상의 분할을 찾을 때 고려할 함수
        - auto : 제곱근으로 함
        - sqrt : 제곱근
        - log2 : 로그
* random_state:int, RandomState instance or None, default=None
    - 추정량의 임의성을 제어함
* ccp_alphanon-negative: float, default=0.0
    - pruning에 사용하는 파라미터
    - 최소비용-복잡성 정리에 사용되는 복잡성 매개변수
    - ccp_alpha보다 작은 비용-복잡성을 가진 서브트리 중 가장 비용-복잡성이 큰 트리를 선택함
    - default로 할 경우 pruning은 수행되지 않음

> sklearn.tree.DecisionTreeRegressor(*, criterion='squared_error', splitter='best', max_depth=None, min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_features=None, random_state=None, max_leaf_nodes=None, min_impurity_decrease=0.0, ccp_alpha=0.0)

In [1]:
import pandas as pd 
df = pd.read_csv('C:/adp/data/kc_house_data.csv')
df 

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors waterfront  view  condition  grade  \
0             1180      5650     1.0   standard     0          3      7   
1             2570      7242     2.0   standard     0          3      7   
2              770     10000     1.0   standard     0          3      6   
3             1960      5000     1.0   standard     0          5      7   
4             1680      8080     1.0   standard     0          3      8   
...            ...       ...     ...        ...   ...        ...    ...   
21608         1530      1131     3.0   standard     0          3      8   
21609         2310      5813     2.0   standard     0          3      8   
21610         1020      1350     2.0   standard     0          3      7   
21611         1600      2388     2.0   standard     0          3      8   
21612         1020      1076     2.0   standard     0          3      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  sqft_living15  \
0            1180              0      1955             0           1340   
1            2170            400      1951          1991           1690   
2             770              0      1933             0           2720   
3            1050            910      1965             0           1360   
4            1680              0      1987             0           1800   
...           ...            ...       ...           ...            ...   
21608        1530              0      2009             0           1530   
21609        2310              0      2014             0           1830   
21610        1020              0      2009             0           1020   
21611        1600              0      2004             0           1410   
21612        1020              0      2008             0           1020   

       sqft_lot15  
0            5650  
1            7639  
2            8062  
3            5000  
4            7503  
...           ...  
21608        1509  
21609        7200  
21610        2007  
21611        1287  
21612        1357  

[21613 rows x 18 columns]

In [ ]:
df = df.drop(['id','date'], axis=1)

In [5]:
X = df.drop('price', axis=1)
y = df['price']

In [6]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   bedrooms       21613 non-null  int64  
 1   bathrooms      21613 non-null  float64
 2   sqft_living    21613 non-null  int64  
 3   sqft_lot       21613 non-null  int64  
 4   floors         21613 non-null  float64
 5   waterfront     21613 non-null  object 
 6   view           21613 non-null  int64  
 7   condition      21613 non-null  int64  
 8   grade          21613 non-null  int64  
 9   sqft_above     21613 non-null  int64  
 10  sqft_basement  21613 non-null  int64  
 11  yr_built       21613 non-null  int64  
 12  yr_renovated   21613 non-null  int64  
 13  sqft_living15  21613 non-null  int64  
 14  sqft_lot15     21613 non-null  int64  
dtypes: float64(2), int64(12), object(1)
memory usage: 2.5+ MB


In [7]:
X = pd.get_dummies(data = X, columns=['waterfront'])

In [8]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3, random_state=2021)

In [12]:
from sklearn.tree import DecisionTreeRegressor
dt_reg = DecisionTreeRegressor()
dt_reg.fit(X_train, y_train)

print(dt_reg.score(X_train, y_train))
print(dt_reg.score(X_test,  y_test))

0.9993379050282394
0.5454481794402503


In [11]:
dt_reg.get_depth()

40

In [13]:
from sklearn.model_selection import GridSearchCV 
param_grid = {'max_depth':range(1,40,2), 'min_samples_split':range(2,40,2)}

model_grid_tree  = GridSearchCV(DecisionTreeRegressor(),param_grid)


In [14]:
help(GridSearchCV)

Help on class GridSearchCV in module sklearn.model_selection._search:

class GridSearchCV(BaseSearchCV)
 |  GridSearchCV(estimator, param_grid, *, scoring=None, n_jobs=None, refit=True, cv=None, verbose=0, pre_dispatch='2*n_jobs', error_score=nan, return_train_score=False)
 |  
 |  Exhaustive search over specified parameter values for an estimator.
 |  
 |  Important members are fit, predict.
 |  
 |  GridSearchCV implements a "fit" and a "score" method.
 |  It also implements "score_samples", "predict", "predict_proba",
 |  "decision_function", "transform" and "inverse_transform" if they are
 |  implemented in the estimator used.
 |  
 |  The parameters of the estimator used to apply these methods are optimized
 |  by cross-validated grid-search over a parameter grid.
 |  
 |  Read more in the :ref:`User Guide <grid_search>`.
 |  
 |  Parameters
 |  ----------
 |  estimator : estimator object
 |      This is assumed to implement the scikit-learn estimator interface.
 |      Either est

In [15]:
model_grid_tree.fit(X_train,y_train)

GridSearchCV(estimator=DecisionTreeRegressor(),
             param_grid={'max_depth': range(1, 40, 2),
                         'min_samples_split': range(2, 40, 2)})

In [11]:
model_grid_tree.best_estimator_

DecisionTreeRegressor(max_depth=7, min_samples_split=16)

In [16]:
model_grid_tree.score(X_train,y_train)

0.7422795151739846

In [17]:
model_grid_tree.score(X_test,y_test)

0.6921699230139526

In [18]:
y_pred = model_grid_tree.predict(X_test)

In [19]:
from sklearn.metrics import mean_squared_error
MSE = mean_squared_error(y_test, y_pred)
MSE

40895468485.00188

In [20]:
import numpy as np
RMSE = np.sqrt(MSE)
RMSE

202226.2804014401

설명 

1. 의사결정나무의 R^2 값은 0.69로 모델이 전체 데이터의 69%를 설명할 수 있다. 
2. RMSE의 값은 202,226 이므로, 집값의 평균적인 차이는 202,226만큼 난다고 할 수 있다. 